core/services/ocr2/plugins/ocr2keeper/evm/mercury_lookup.go

```diff
@@ -0,0 +1,343 @@
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/patrickmn/go-cache"
+	"github.com/smartcontractkit/ocr2keepers/pkg/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/keeper_registry_wrapper2_0"
+)
+
+type MercuryLookup struct {
+	feedLabel  string
+	feeds      []string
+	queryLabel string
+	query      *big.Int
+	extraData  []byte
+}
+
+// MercuryMultiResponse TODO guessing on this
+type MercuryMultiResponse struct {
+	ChainlinkBlobs []string `json:"chainlinkBlob"`
+}
+
+type MercuryResponse struct {
+	ChainlinkBlob string `json:"chainlinkBlob"`
+}
+
+type MercuryBytes struct {
+	Index int
+	Error error
+	Bytes []byte
+}
+
+const (
+	RetryDelay   = 500 * time.Millisecond
+	TotalAttempt = 3
+)
+
+// mercuryLookup looks through check upkeep results looking for any that need off chain lookup
+func (r *EvmRegistry) mercuryLookup(ctx context.Context, upkeepResults []types.UpkeepResult) ([]types.UpkeepResult, error) {
+	// return error only if there are errors which stops the process
+	// don't surface Mercury API errors to plugin bc MercuryLookup process should be self-contained
+	for i := range upkeepResults {
+		// if its another reason continue/skip
+		if upkeepResults[i].FailureReason != UPKEEP_FAILURE_REASON_TARGET_CHECK_REVERTED {
+			continue
+		}
+
+		block, upkeepId, err := blockAndIdFromKey(upkeepResults[i].Key)
+		if err != nil {
+			r.lggr.Error("[MercuryLookup] error getting block and upkeep id:", err)
+			return nil, err
+		}
+
+		// checking if this upkeep is in cooldown from api errors
+		_, onIce := r.mercury.cooldownCache.Get(upkeepId.String())
+		if onIce {
+			r.lggr.Infof("[MercuryLookup] cooldown Skipping UpkeepId %s at block %s\n", upkeepId.String(), block.String())
+			continue
+		}
+
+		// if it doesn't decode to the mercury custom error continue/skip
+		mercuryLookup, err := r.decodeMercuryLookup(upkeepResults[i].PerformData)
+		if err != nil {
+			upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_MERCURY_LOOKUP_ERROR
+			r.lggr.Debugf("[MercuryLookup] decodeMercuryLookup: %v", err)
+			return nil, err
+		}
+
+		opts, err := r.buildCallOpts(ctx, block)
+		if err != nil {
+			r.lggr.Errorf("[MercuryLookup] buildCallOpts: %v", err)
+			return nil, err
+		}
+		// need upkeep info for offchainConfig and to hit callback
+		upkeepInfo, err := r.getUpkeepInfo(upkeepId, opts)
+		if err != nil {
+			r.lggr.Errorf("[MercuryLookup] GetUpkeep: %v", err)
+			return nil, err
+		}
+
+		// 	do the http request
+		values, err := r.doRequest(ctx, mercuryLookup, upkeepId)
+		if err != nil {
+			upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_MERCURY_LOOKUP_ERROR
+			r.lggr.Errorf("[MercuryLookup] doRequest: %v", err)
+			continue
+		}
+
+		needed, performData, err := r.mercuryLookupCallback(ctx, mercuryLookup, values, upkeepInfo, opts)
+		if err != nil {
+			upkeepResults[i].FailureReason = UPKEEP_FAILURE_REASON_MERCURY_LOOKUP_ERROR
```

Comment:
can we differentiate the error here from line 104

Reply:
i believe a few more failure reasons were added to differentiate why upkeeps fail here. but these are not part of the automation compatible interface. since we will treat mercury lookup a part of check upkeep. i changed them to `UPKEEP_FAILURE_REASON_TARGET_CHECK_REVERTED`